### PR TITLE
ninja: Add re2c build dependency

### LINKED ISSUE #33127: Installation issue: ninja
### Steps to reproduce the issue

```console
$ spack spec -l ninja@1.11.0                                                                                           
Input spec
--------------------------------
ninja@1.11.0

Concretized
--------------------------------
twjnp4e  ninja@1.11.0%gcc@4.8.5 arch=linux-centos7-haswell
u3amfhc      ^python@3.9.13%gcc@4.8.5+bz2+ctypes+dbm~debug+libxml2+lzma~nis~optimizations+pic+pyexpat+pythoncmd+readline+shared+sqlite3+ssl~tix~tkinter~ucs4+uuid+zlib patches=0d98e93,4c24573,f2fd060 arch=linux-centos7-haswell
4rdc3sz          ^bzip2@1.0.8%gcc@4.8.5~debug~pic+shared arch=linux-centos7-haswell
jbmktfe              ^diffutils@3.8%gcc@4.8.5 arch=linux-centos7-haswell
rtwsiol                  ^libiconv@1.16%gcc@4.8.5 libs=shared,static arch=linux-centos7-haswell
npj2eln          ^expat@2.4.8%gcc@4.8.5+libbsd arch=linux-centos7-haswell
jo5ve7e              ^libbsd@0.11.5%gcc@4.8.5 arch=linux-centos7-haswell
itekqgt                  ^libmd@1.0.4%gcc@4.8.5 arch=linux-centos7-haswell
5wdqfdk          ^gdbm@1.19%gcc@4.8.5 arch=linux-centos7-haswell
4zm4g6g              ^readline@8.1.2%gcc@4.8.5 arch=linux-centos7-haswell
mbqxdtu                  ^ncurses@6.3%gcc@4.8.5~symlinks+termlib abi=none arch=linux-centos7-haswell
pirjvlv                      ^pkgconf@1.8.0%gcc@4.8.5 arch=linux-centos7-haswell
iiumocu          ^gettext@0.21%gcc@4.8.5+bzip2+curses+git~libunistring+libxml2+tar+xz arch=linux-centos7-haswell
pprckta              ^libxml2@2.10.1%gcc@4.8.5~python arch=linux-centos7-haswell
3huq7ap                  ^xz@5.2.5%gcc@4.8.5~pic libs=shared,static arch=linux-centos7-haswell
o24fh5i                  ^zlib@1.2.12%gcc@4.8.5+optimize+pic+shared patches=0d38234 arch=linux-centos7-haswell
53rbfhq              ^tar@1.34%gcc@4.8.5 zip=pigz arch=linux-centos7-haswell
ccdc54s                  ^pigz@2.7%gcc@4.8.5 arch=linux-centos7-haswell
iyj2tb4                  ^zstd@1.5.2%gcc@4.8.5+programs compression=none libs=shared,static arch=linux-centos7-haswell
qgvdaua          ^libffi@3.4.2%gcc@4.8.5 arch=linux-centos7-haswell
5e3p4o6          ^openssl@1.1.1q%gcc@4.8.5~docs~shared certs=mozilla patches=3fdcf2d arch=linux-centos7-haswell
fhqdsbp              ^ca-certificates-mozilla@2022-07-19%gcc@4.8.5 arch=linux-centos7-haswell
ojqfu5a              ^perl@5.34.1%gcc@4.8.5+cpanm+shared+threads arch=linux-centos7-haswell
twsa6dw                  ^berkeley-db@18.1.40%gcc@4.8.5+cxx~docs+stl patches=26090f4,b231fcc arch=linux-centos7-haswell
4xxmxqp          ^sqlite@3.39.2%gcc@4.8.5+column_metadata+dynamic_extensions+fts~functions+rtree arch=linux-centos7-haswell
jl2hrnb          ^util-linux-uuid@2.36.2%gcc@4.8.5 arch=linux-centos7-haswell

```

### Error message

<details><summary>Error message</summary><pre>
==> Installing ninja-1.11.0-twjnp4ekwdrwtqak6hp5hj6v7fmli3uc
==> No binary for ninja-1.11.0-twjnp4ekwdrwtqak6hp5hj6v7fmli3uc found: installing from source
==> Using cached archive: /rwthfs/rz/cluster/work/sg805090/local/spack/var/spack/cache/_source-cache/archive/3c/3c6ba2e66400fe3f1ae83deb4b235faf3137ec20bd5b08c29bfc368db143e4c6.tar.gz
==> No patches needed for ninja
==> ninja: Executing phase: 'configure'
==> Error: ProcessError: Command exited with status 1:
    '/rwthfs/rz/cluster/work/sg805090/local/spack/opt/spack/linux-centos7-haswell/gcc-4.8.5/python-3.9.13-u3amfhc3iny6rjjzspzg26meqf4rhso2/bin/python3.9' 'configure.py' '--bootstrap'

1 error found in build log:
  >> 3    Error in argument 4, char 1: no argument for option -
     4    usage: re2c [-bcdDefFghirsuvVwx18] [-o of] [-t th] file
     5    
     6    -? -h  --help           Display this info.
     7    
     8    -b     --bit-vectors    Implies -s. Use bit vectors as well in the attempt to
     9                            coax better code out of the compiler. Most useful for

See build log for details:
  /tmp/sg805090/login18-2_125808/spack-stage/spack-stage-ninja-1.11.0-twjnp4ekwdrwtqak6hp5hj6v7fmli3uc/spack-build-out.txt
</pre></details>

### Information on your system

* **Spack:** 0.19.0.dev0
* **Python:** 3.6.8
* **Platform:** linux-centos7-skylake_avx512
* **Concretizer:** clingo

### Additional information

[spack-build-out.txt](https://github.com/spack/spack/files/9741139/spack-build-out.txt)
[spack-build-env.txt](https://github.com/spack/spack/files/9741140/spack-build-env.txt)
[spack-build-env-mods.txt](https://github.com/spack/spack/files/9741141/spack-build-env-mods.txt)

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

### PR DESCRIPTION
**Pull request by picsel2**
Fixes spack/spack#33127

### CODE REVIEW

File: var/spack/repos/builtin/packages/ninja/package.py [hunk 1]
```diff
@@ -35,6 +35,7 @@ class Ninja(Package):
     version("1.6.0", sha256="b43e88fb068fe4d92a3dfd9eb4d19755dae5c33415db2e9b7b61b4659009cde7")
 
     depends_on("python", type="build")
+    depends_on("re2c", type="build")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/ninja/package.py]:
```suggestion
    depends_on("re2c@0.11.3:", type="build")
```

**Comment by adamjstewart** [comment]
The `re2c` package uses ninja to build on Windows, so this will introduce a circular dependency.

**Comment by picsel2** [comment]
That sounds awful.

I'm using spack for the first time. I have no clue how such a circular dependency is supposed to be handled.

**Comment by adamjstewart** [comment]
Maybe @johnwparent can comment on this one.

**Review by haampie — APPROVED** [review]
Whatever the Windows folks are doing to break circular deps, this change should get in.

**Comment by johnwparent** [comment]
@adamjstewart While re2c does depend on Ninja to build the way the package is structured, no dependency is actually created, so it will result in a build error rather than a circular dep. Not that that's really much better. 

Further,
To me it doesn't look like Ninja actually _needs_ `re2c` to build, and can build just fine without it (and does, based on some local testing without re2c in my PATH), so long as an outdated version of re2c is not present on the PATH. In the issue it was indicated a system version from outside Spack was picked up. It seems like [changes to re2c](https://github.com/ninja-build/ninja/issues/2192) resulted in this issue, rather than Ninja suddenly adding a hard requirement. From where I'm standing this solution is treating a symptom and not a cause. My proposed solution would then be to make this a variant of the Ninja package, that way we avoid the circular dependency, but still allow for a Spack derived re2c to be picked up as needed.

For example:

```
variant("re2c", default=True)

depends_on("re2c", when="+re2c")
```

and then in re2c:

`depends_on("ninja~re2c")`

**Review by bernhardkaindl — APPROVED** [review]
I agree with the approval from haampie and disagree with the assessment of [johnwparent](https://github.com/johnwparent) due to this comment: https://github.com/ninja-build/ninja/issues/1703#issuecomment-572226503
```py
ninja -v
[1/4] cd /home/crondog/code/ninja/build && /usr/bin/re2c -b -i
```
This shows that re2c is used by ninja when building other projects, so it should still stay a dependency as other recipes may depend on it being present. Merging.

PS: The build of re2c on windows using ninja is not an issue because it just assumes that ninja is present but does not declare a dependency on it. This way, the re2c recipe this avoids problems for the concretizer by not creating a circular dependency chain.

**Pull request merged**